File: 494-target-sum/target-sum.cpp
class Solution {
private:
    int OFFSET = 1000;
    int dp[22][2004];
public:
    Solution(){
        memset(dp, -1, sizeof(dp));
    }
    int solve(vector<int> &nums, int target, int n, int sum){
        if (n < 0){
            if (sum == target) return 1;
            return 0;
        }
        
        if (dp[n][OFFSET+sum] != -1) return dp[n][OFFSET+sum];

        int add = solve(nums, target, n-1, sum + nums[n]);
        int sub = solve(nums, target, n-1, sum - nums[n]);

        dp[n][OFFSET+sum] = add + sub;
        
        return dp[n][OFFSET+sum];
    }
    int findTargetSumWays(vector<int>& nums, int target) {
        int sum = 0;
        int n = nums.size()-1;
        int cnt = solve(nums, target, n, sum);
        return cnt;
    }
};

File: 494-target-sum/target-sum.go
var dp = make([][]int, 22)
const OFFSET = 1001
func solve(nums []int, target, sum, idx int) int {
    if idx >= len(nums) {
        if target == sum {
            return 1
        }
        return 0
    }

    if dp[idx][OFFSET+sum] != -1 {
        return dp[idx][OFFSET+sum]
    }

    add := solve(nums, target, sum + nums[idx], idx+1)
    sub := solve(nums, target, sum - nums[idx], idx+1)

    dp[idx][OFFSET+sum] = add + sub
    return dp[idx][OFFSET+sum]
}

func findTargetSumWays(nums []int, target int) int {
    initDP()
    return solve(nums, target, 0, 0)
}

func initDP() {
    for i := range dp {
        dp[i] = make([]int, 2200)
        for j := range dp[i] {
            dp[i][j] = -1
        }
    }
}